hudi-client/hudi-client-common/src/main/java/org/apache/hudi/io/HoodieMergeHandle.java

```diff
@@ -310,8 +310,19 @@ private boolean writeRecord(HoodieRecord<T> newRecord, Option<HoodieRecord> comb
     }
     try {
       if (combineRecord.isPresent() && !combineRecord.get().isDelete(schema, config.getProps()) && !isDelete) {
-        writeToFile(newRecord.getKey(), combineRecord.get(), schema, prop, preserveMetadata && useWriterSchemaForCompaction);
-        recordsWritten++;
+        // Last-minute check.
+        Pair<Boolean, Boolean> decision = recordMerger.shouldFlush(combineRecord.get(), schema);
+
+        if (decision.getLeft()) { // CASE (1): Flush the merged record.
+          writeToFile(newRecord.getKey(), combineRecord.get(), schema, prop, preserveMetadata && useWriterSchemaForCompaction);
+          recordsWritten++;
+        } else {
+          if (decision.getRight()) { // CASE (2): Flush the old record only.
+            return false;
```

Comment:
Thanks for the fix, we might need some unit tests for this new merger interface, with three different combinations of the returned pair: (true, _), (false, true), (false, false).

Reply:
@danny0405 since the returned value is only used in mergehandle, we should add some tests for mergehandle. Let me find how to do that.

Reply:
Added functional tests to test the behavior of this function.